src/FieldWarning/Assets/Units/Component/Armor/ArmorComponent.cs

```diff
@@ -37,18 +38,18 @@ class ArmorComponent
         /// An array of ArmorAttributes to store armor values
         /// on front / side / rear/ top respectively
         /// </summary>
-        public UnitData.ArmorAttributes[] ArmorData = new UnitData.ArmorAttributes[4];
+        public ArmorAttributes[] ArmorData = new ArmorAttributes[4];
 
-        public ArmorComponent(
-            MovementComponent unit,
-            PlatoonBehaviour platoon,
-            UnitData.ArmorAttributes[] armorData,
-            HealthComponent healthComponent)
+        private void Awake()
         {
-            Unit = unit;
-            _platoon = platoon;
-            ArmorData = armorData;
-            _healthComponent = healthComponent;
+            ArmorData = gameObject.GetComponent<DataComponent>().ArmorData;
+            Unit = gameObject.GetComponent<MovementComponent>();
+            _healthComponent = gameObject.GetComponent<HealthComponent>();
+        }
```

Comment:
This is one of the big disadvantages of MonoBehaviours, when you have a constructor you can look up the references to that constructor and figure out where your data came from. When you just have GetComponent<T>() calls, it is a bit hard to chase down who created those components for you originally and what guarantees that they did so before your Awake() method was called.

The best middle ground I've seen so far is to define a WakeUp( ..constructorArgs.. ) method that sets your fields as a constructor would and then enables your component (awaking it). That gives us both the explicit dependency passing of constructors and the automated serialization from MonoBehaviour. 

No need to change it now, just food for thought..

Reply:
Yeah, the problem I ran into with the non-MonoBehaviour components was that they're not serialized by Unity, so the data wasn't persisting through the prototype creation -> cloning at purchase-time workflow.

But also, I found it to be much cleaner to have the GameObject itself as a sort of single source of truth for all of the properties and methods that the various components need access to for interoperability. Before this, some of the components were getting, e.g., the Data by going through the movement component; some were going through the dispatcher; some were getting it directly, etc. This way it's more consistent between all of the components, and you also don't have a bunch of different constructors taking a bunch of different parameter sets. IIRC, the dispatcher was the only oddball that needed to be initialized after the prototyping stage (since it doesn't come into play until after the unit actually spawns but a lot of the components needed access to it), so manually passing it on to the components that needed it seemed like the lesser evil.

I do really like the idea of that WakeUp pattern though; I remember seeing a method with that name in a few places but I didn't really look too closely at them. I can go ahead and replace those Initialize methods with WakeUp calls either tonight or tomorrow.

Thanks for taking the time to review this btw! I'm pretty naive to Unity/C# so I really appreciate the constructive criticism.

Reply:
Oh, also:
> when you have a constructor you can look up the references to that constructor and figure out where your data came from. When you just have GetComponent() calls, it is a bit hard to chase down who created those components for you originally

I think this is mitigated by the fact that they're all created in the same place (the UnitFitter class), but I didn't really consider that somebody who didn't already know that might have some headache trying to track it down. Do you think some comments or documentation somewhere might resolve that?

Reply:
> But also, I found it to be much cleaner to have the GameObject itself as a sort of single source of truth for all of the properties and methods that the various components need access to for interoperability.

I agree that what you did with the data component is a big improvement in readability. Originally all of the functionality was in one `UnitBehaviour `class, which I have been trying to split up, but the components didn't end up entirely independent from each other.

> Do you think some comments or documentation somewhere might resolve that?

I don't think it's a huge issue in the first place, the dispatcher is central enough that even new readers of the code should figure it out relatively quickly. I just figured I'd point it out as a potential drawback of doing too much stuff in Awake().